File: peek_plugin_chat/plugin-module/_private/tuples/ChatTuple.ts
import {addTupleType, Tuple} from "@synerty/vortexjs";
import {chatTuplePrefix} from "../PluginNames";
import {ChatUserTuple} from "./ChatUserTuple";
import {MessageTuple} from "./MessageTuple";


@addTupleType
export class ChatTuple extends Tuple {
    public static readonly tupleName = chatTuplePrefix + "ChatTuple";

    //  Description of date1
    id: number;

    // Message details
    hasUnreads: boolean;
    lastActivity: Date;

    messages: MessageTuple[] = [];
    users: ChatUserTuple[] = [];

    constructor() {
        super(ChatTuple.tupleName)
    }
}

File: peek_plugin_chat/_private/storage/ChatTuple.py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy.orm import relationship
from sqlalchemy.sql.sqltypes import DateTime, Boolean

from peek_plugin_chat._private.PluginNames import chatTuplePrefix
from peek_plugin_chat._private.storage.DeclarativeBase import DeclarativeBase
from vortex.Tuple import Tuple, addTupleType, TupleField


@addTupleType
class ChatTuple(Tuple, DeclarativeBase):
    __tupleType__ = chatTuplePrefix + 'ChatTuple'
    __tablename__ = 'ChatTuple'

    # This will include the users relationship when serialising the data
    __fieldNames__ = ["users"]

    id = Column(Integer, primary_key=True, autoincrement=True)

    # Message details
    hasUnreads = Column(Boolean, nullable=False)
    lastActivity = Column(DateTime, nullable=False)

    # Use a TupleField instead of a relationship so we can decide when it will
    # include the messages or not.
    messages = TupleField()  # relationship("Message.chatId")

    users = relationship("ChatUserTuple", lazy='joined')
